pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,539 @@
+			})
+			.map(|x| x.clone())
+			.collect::<Vec<_>>();
+
+		if res.is_empty() {
+			return Err(Error::<T>::NotEnoughRoutersForDomain.into());
+		}
+
+		Ok(res)
+	}
+
+	/// Calculates and returns the proof count required for processing one
+	/// inbound message.
+	pub(crate) fn get_expected_proof_count(
+		router_ids: &Vec<T::RouterId>,
+	) -> Result<u32, DispatchError> {
+		let expected_proof_count = router_ids
+			.len()
+			.ensure_sub(1)
+			.map_err(|_| Error::<T>::NotEnoughRoutersForDomain)?;
+
+		Ok(expected_proof_count.saturated_into())
+	}
+
+	/// Gets the message proof for a message.
+	pub(crate) fn get_message_proof(message: T::Message) -> Proof {
+		match message.get_message_proof() {
+			None => message
+				.to_message_proof()
+				.get_message_proof()
+				.expect("message proof ensured by 'to_message_proof'"),
+			Some(proof) => proof,
+		}
+	}
+
+	/// Upserts an inbound entry for a particular message, increasing the
+	/// relevant counts accordingly.
+	pub(crate) fn upsert_pending_entry(
+		message_proof: Proof,
+		router_id: T::RouterId,
+		new_inbound_entry: InboundEntry<T>,
+	) -> DispatchResult {
+		PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+			match storage_entry {
+				None => {
+					*storage_entry = Some(new_inbound_entry);
+
+					Ok::<(), DispatchError>(())
+				}
+				Some(stored_inbound_entry) => {
+					stored_inbound_entry.pre_dispatch_update(new_inbound_entry)
+				}
+			}
+		})
+	}
+
+	/// Creates, validates and upserts the inbound entry.
+	fn validate_and_upsert_pending_entries(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message: T::Message,
+		message_proof: Proof,
+		router_id: T::RouterId,
+	) -> DispatchResult {
+		let inbound_entry: InboundEntry<T> = (inbound_processing_info, message).into();
+
+		inbound_entry.validate(&inbound_processing_info, &router_id)?;
+
+		Self::upsert_pending_entry(message_proof, router_id, inbound_entry)?;
+
+		Ok(())
+	}
+
+	/// Checks if the number of proofs required for executing one message
+	/// were received, and if so, decreases the counts accordingly and executes
+	/// the message.
+	pub(crate) fn execute_if_requirements_are_met(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message_proof: Proof,
+	) -> DispatchResult {
+		let mut message = None;
+		let mut votes = 0;
+
+		for router_id in &inbound_processing_info.router_ids {
+			match PendingInboundEntries::<T>::get(message_proof, router_id) {
+				// We expected one InboundEntry for each router, if that's not the case,
+				// we can return.
+				None => return Ok(()),
+				Some(stored_inbound_entry) => match stored_inbound_entry {
+					InboundEntry::Message(message_entry) => message = Some(message_entry.message),
+					InboundEntry::Proof(proof_entry) => {
+						if proof_entry
+							.has_valid_vote_for_session(inbound_processing_info.current_session_id)
+						{
+							votes.ensure_add_assign(1)?;
+						}
+					}
```

Comment:
Nit/food for thought: Could pull out the if condition, but not sure if it improves readability since then we have to handle the remainder case.
```suggestion
					InboundEntry::Proof(proof_entry) if proof_entry .has_valid_vote_for_session(inbound_processing_info.current_session_id) => {
						votes.ensure_add_assign(1)?;
					},
					_ => {},
```

Reply:
Changed. Thank you!